sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java

```diff
@@ -0,0 +1,474 @@
+ *       .withDatabase("default") //optional, assumes default if none specified
+ *       .withFilter(filterString) //optional,
+ *       should be specified if the table is partitioned
+ * }</pre>
+ *
+ *
+ *
+ * <h3>Writing to Hive</h3>
+ *
+ * <p>Hive sink supports writing of HCatRecord to a Hive instance.</p>
+ *
+ * <p>To configure a Hive sink, you must specify a metastoreUri and a table name.
+ * Other optional parameters are database, partition & batchsize
+ * For instance:</p>
+ *
+ * <pre>{@code
+ *
+ * pipeline
+
+ *   .apply(...)
+ *   .apply(HiveIO.write()
+ *       .withMetastoreUri("thrift://hadoop-clust-0118-m:9083") //mandatory
+ *       .withTable("employee") //mandatory
+ *       .withDatabase("default") //optional, assumes default if none specified
+ *       .withFilter(partitionValues) //optional,
+ *       should be specified if the table is partitioned
+ *       .withBatchSize(1024L)) //optional,
+ *       assumes a default batch size of 1024 if none specified
+ * }</pre>
+ *
+
+ */
+@Experimental
+public class HiveIO {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HiveIO.class);
+
+  /** Write data to Hive. */
+  public static Write write() {
+    return new AutoValue_HiveIO_Write.Builder().setBatchSize(1024L).build();
+  }
+
+  /** Read data from Hive. */
+  public static Read read() {
+    return new AutoValue_HiveIO_Read.Builder().build();
+  }
+
+  private HiveIO() {
+  }
+
+
+  /**
+   * A {@link PTransform} to read data from Hive.
+   */
+  @AutoValue
+  public abstract static class Read extends PTransform<PBegin, PCollection<HCatRecord>> {
+    @Nullable abstract String getMetastoreUri();
+    @Nullable abstract String getDatabase();
+    @Nullable abstract String getTable();
+    @Nullable abstract String getFilter();
+    @Nullable abstract Coder<HCatRecord> getCoder();
+    @Nullable abstract ReaderContext getContext();
+    @Nullable abstract Integer getSplitId();
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setMetastoreUri(String metastoreUri);
+      abstract Builder setDatabase(String database);
+      abstract Builder setTable(String table);
+      abstract Builder setFilter(String filter);
+      abstract Builder setCoder(Coder<HCatRecord> coder);
+      abstract Builder setSplitId(Integer splitId);
+      abstract Builder setContext(ReaderContext context);
+      abstract Read build();
+    }
+
+    public Read withMetastoreUri(String metastoreUri) {
+      return toBuilder().setMetastoreUri(metastoreUri).build();
+    }
+
+    public Read withDatabase(String database) {
+      return toBuilder().setDatabase(database).build();
+    }
+
+    public Read withTable(String table) {
+      return toBuilder().setTable(table).build();
+    }
+
+    public Read withFilter(String filter) {
+      return toBuilder().setFilter(filter).build();
+    }
+
+    protected Read withSplitId(Integer splitId) {
```

Comment:
Why not int - can it be null? If yes, mark it `@Nullable` and document what does it mean if you supply null